package/MDAnalysis/analysis/Aggregation_size.py

```diff
@@ -0,0 +1,364 @@
+runnig the code.
+2- The periodic boundary should be eliminated before running the code. We 
+recommend doing -pbc whole and -pbc cluster using trjconv keyword (available)
+in GROMCAS or any similar procedure. 
+3- If you have more than one type of molecule and you want to measure the
+aggregation without considering the type of molecule the code will works
+perfectly for "Closest" and "COM" Types while for the "Specific" the results
+will not be correct. The reasone is that you cannot tell atom_1 and atom_2 is
+related to which residue!
+4- If you have more than one type of molecule and you want to measure
+aggregation for each type of molecule separately, you can keep only the
+target molecule in the box and run the code multiple times for different
+molecule type.
+5- You will get an error if there is no aggregation in the box or if the 
+aggregation distance choose too short. 
+"""
+
+
+def aggregate(k, tss, counter, m, Limit):
+    """
+    This is a subfunction that enables the code to find all 
+    connected molecules in a cluster.
+    k is a connected molecule to the main molecule in the first level.
+    tss is a time frame.
+    counter is a step.
+    m provides a suitable off_set to match the number of molecules.
+    Limit is the cut_off provided by the user.
+    """
+    import numpy as np
+    bool_sector1 = globals()["d" + str(k)][:, [tss]] < Limit
+    Coordinates1 = globals()["d" + str(k)][:, [tss]][bool_sector1]
+    # This "if" will check if there is any uninvestigated molecule left
+    if len(Coordinates1) != 0:
+        index1 = np.where(bool_sector1)
+        listOfCoordinates1 = list(zip(index1[0], index1[1]))
+        connect = ([j[0] for j in listOfCoordinates1]
+                   + globals()["connected" + str(counter)][m] + 1)
+        rt = np.array(connect)
+    else:
+        rt = []
+    return rt
+
+
+def Aggregation_size(gro_file, xtc_file, number_of_molecules, Type, monomer="No", Gyr_calc="No", cut_off=False,
+                        atom_1=None, atom_2=None):
+    """
+    1- Import the needed module
+    2- Calculate the distance between molecules based on the intended criteria
+    3- Find the aggregates by checking only one time each using the
+    "aggregate" subfunction (to be efficient)
+    4- Check and merge two aggregates as one aggregate if the distance
+    between any element of them is less than cut_off
+    5- Eliminate monomers (optional)
+    6- measure average gyration radius (optional)
+    7- measure the number of aggregates with different size
+    """
+    import warnings
+    import numpy as np
+    import MDAnalysis as mda
+    from MDAnalysis.lib import distances
+    u = mda.Universe(gro_file, xtc_file)
+    molnum = number_of_molecules
+    Limit = cut_off
+    i = 0
+    for it in reversed(range(1, molnum)):
+        i += 1
+        globals()["d" + str(i)] = np.zeros(
+            it * len(u.trajectory)).reshape(it, len(u.trajectory))
+    if Type == "Closest":
```

Comment:
I am now using the _single_frame function in the AnalysisBase class. So, I need to run through the frames more than once, I am wondering if I need to consider _single_frame for all loops which are going through frames? if yes how can I accommodate that?

Reply:
If I recall correctly, your initial logic was to first compute the distance matrix for all the frames, then cluster the the particles into aggregates for all the frames. With `_single_frame` I would recommand to change you logic a little bit. Instead of looping twice over the frames, I would, for each frame, calculate the distance matrix and do the clustering. This is possible because you do not need the information from the other frames when you compute the aggregate for one given frame.

By doing all the frames independently from each other, you avoid looping multiple times over the frames, but you also avoid storing large amount of data that you do not need. By building all the distance matrices at once, you have to store a large amount of data (n*n*f, with n the number of atoms, and f the number of frames) which limit the size of the trajectories you can work with. By doing the frames separately, you only have one distance matrix in memory at a time.